src/docfx/build/toc/parser/MarkdownTocMarkup.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Markdig.Extensions.Yaml;
+using Markdig.Syntax;
+using Markdig.Syntax.Inlines;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class MarkdownTocMarkup
+    {
+        private static readonly HashSet<Type> s_blockWhiteList = new HashSet<Type> { typeof(HeadingBlock) /*header*/, typeof(YamlFrontMatterBlock) /*yaml header*/, typeof(HtmlBlock) /*comment*/ };
+
+        public static (List<Error> errors, TableOfContentsInputModel model) LoadMdTocModel(string tocContent, string filePath)
+        {
+            var errors = new List<Error>();
+            var headingBlocks = new List<HeadingBlock>();
+
+            var (ast, result) = Markup.Parse(tocContent);
+            errors.AddRange(result.Errors);
+            foreach (var block in ast)
+            {
+                if (!s_blockWhiteList.Contains(block.GetType()))
+                {
+                    errors.Add(Errors.InvalidTocSyntax(new Range(block.Line, block.Column), filePath, tocContent.Substring(block.Span.Start, block.Span.Length)));
+                }
+
+                if (block is HeadingBlock headingBlock)
+                {
+                    headingBlocks.Add(headingBlock);
+                }
+            }
+
+            var tocModel = new TableOfContentsInputModel
+            {
+                Metadata = result.Metadata,
+            };
+
+            try
+            {
+                tocModel.Items = ConvertTo(filePath, headingBlocks.ToArray(), errors).children;
+            }
+            catch (DocfxException ex)
+            {
+                errors.Add(ex.Error);
+            }
+
+            return (errors, tocModel);
+        }
+
+        private static (List<TableOfContentsInputItem> children, int count) ConvertTo(string filePath, HeadingBlock[] headingBlocks, List<Error> errors, int startIndex = 0)
+        {
+            if (headingBlocks.Length == 0)
+            {
+                return (new List<TableOfContentsInputItem>(), 0);
+            }
+
+            Debug.Assert(startIndex < headingBlocks.Length);
```

Comment:
can a stack make this algorithm simpler?

Reply:
maybe, but I like my way, more easy for reading and thinking 

---
In reply to: [212534685](https://github.com/dotnet/docfx/pull/3269#discussion_r212534685) [](ancestors = 212534685)